### PR TITLE
Fix/maxlength

### LINKED ISSUE #28733: Input: maxLength  can't constraint the value set by script
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/input-maxlength-msh6i)

### Steps to reproduce
- refresh the frame
- wait 1s 

### What is expected?
value.length can not greater than maxlength

### What is actually happening?
the Input is set over the maxLength value by default,  textarea has been constrainted by hand
the origin element also has the same problem, but it's designed 
we must constraint it by hand since the maxLength only detect the user input(opposed to the way set by script )

| Environment | Info |
|---|---|
| antd | 4.10.0 |
| React | 16.3.1 |
| System | mac |
| Browser | chrome |

### PR DESCRIPTION
**Pull request by HomyeeKing**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [x] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

fix #28733
### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    Fix Input: maxLength only constraint user edit not the way set by script       |
| 🇨🇳 Chinese |      修复 Input 的maxLength 只限制用户输入 没有限制动态设置的value长度的问题    |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by HomyeeKing** [comment]
also fix `Input.TextArea` can't prevent input when maxLength is 0

File: components/input/Input.tsx [hunk 1]
```diff
@@ -330,10 +332,15 @@ class Input extends React.Component<InputProps, InputState> {
...
     const { value, focused } = this.state;
-    const { prefixCls: customizePrefixCls, bordered = true } = this.props;
+    const { prefixCls: customizePrefixCls, bordered = true, maxLength } = this.props;
     const prefixCls = getPrefixCls('input', customizePrefixCls);
     this.direction = direction;
 
+    let val = fixControlledValue(value) as string;
+    // fix: maxlength can't constraint the value set by script, so we need to do it by hand
+    const hasMaxLength = Number(maxLength) >= 0;
+    val = hasMaxLength ? [...val].slice(0, maxLength).join('') : val;
```
Comment (zombieJ) [inline components/input/Input.tsx:342]: Abstract into a general method since these code is duplicated.

**Comment by afc163** [comment]
I stick to https://github.com/ant-design/ant-design/issues/28733#issuecomment-755315065

**Comment by HomyeeKing** [comment]
@afc163  
look at this 
> // fix #27612 将value转为数组进行截取，解决 '😂'.length === 2 等emoji表情导致的截取乱码的问题

also  [this fix](https://github.com/ant-design/ant-design/pull/28734/files#diff-e38236283d4b2ff8417dfed931e619e2652dffcf55a9d22125fc8d2c8c16e089L112) has a problem that doesn't consider the `maxLength = 0` case

and we can still skip the form validation if we set the value more than maxLength which seems wierd

File: components/input/Input.tsx [hunk 2]
```diff
@@ -61,6 +62,14 @@ export function fixControlledValue<T>(value: T) {
   return value;
 }
 
+export const hasMaxLength = (val: number) => val >= 0;
+// fix #27612 将value转为数组进行截取，解决 '😂'.length === 2 等emoji表情导致的截取乱码的问题
+// fix:#28733 maxlength can't constraint the value set by script, so we need to do it by hand
+export function dealWithMaxLength<T>(maxLength: number, value: T) {
```
Comment (zombieJ) [inline components/input/Input.tsx:68]: `truncateValue`

**Comment by HomyeeKing** [comment]
[fix showCount](https://codesandbox.io/s/fix-shoucount-9pj83?file=/index.js)

**Comment by HomyeeKing** [comment]
hey , is there any suggestions about this

**Comment by zombieJ** [comment]
https://github.com/ant-design/ant-design/issues/28733#issuecomment-758542780

> Since we already handle the [special case](https://github.com/ant-design/ant-design/blob/6bcb9ba7c4561b5b1b2c1c874961475f9e141166/components/input/TextArea.tsx#L114), I think we may handle this all?

cc @afc163

**Comment by afc163** [comment]
We may encounter issue like https://github.com/ant-design/ant-design/issues/28940

**Comment by zombieJ** [comment]
To avoid breaking change. Let move this into feature branch for minor version.

**Comment by HomyeeKing** [comment]
> We may encounter issue like #28940

it's also been fix in this commit,  since that's a `native bug` I think

**Comment by HomyeeKing** [comment]
> We may encounter issue like #28940

let me give it a try and see what I can do👀

File: components/input/Input.tsx [hunk 3]
```diff
@@ -179,7 +189,16 @@ class Input extends React.Component<InputProps, InputState> {
   }
 
   componentDidMount() {
+    const { inputLock, value } = this.state;
+    const { maxLength } = this.props;
     this.clearPasswordValueAttribute();
+    setTimeout(() => {
+      if (!inputLock && hasMaxLength(maxLength) && maxLength === 5) {
+        this.setState({ value: truncateValue(Number(maxLength), value) }, () => {
+          console.log(this.state);
```
Comment (HomyeeKing) [inline components/input/Input.tsx:198]: I really really really have no idea why this `state` not  be updated

**Comment by HomyeeKing** [comment]
manage class component's `props` and `state` is really really a nightmare for a react newbie😱😱

**Comment by HomyeeKing** [comment]
@zombieJ  
hey , since I am working on this for a few days, I have some thoughts about this
Firstly, the `Input` and `TextArea` is working well now  as we expect , but with the effect is the some `Input` test case failed
because of  props and state manage I think

As a react newbie and based on what I've done, it would be better to reduce useless work and  make this sensible
so  I have a solution or suggestion:

- Reserve the `TextArea` fix since it works well and no test case fail effect
- Rewrite the `Input` with hooks, then we make the fix after that

before that, I'd like you to review my code , both `Input` and `TextArea` 

what do you think about?

File: components/input/TextArea.tsx [hunk 4]
```diff
@@ -68,10 +77,23 @@ const TextArea = React.forwardRef<TextAreaRef, TextAreaProps>(
...
+        handleSetValue(e.target.value);
+      }
       resolveOnChange(innerRef.current as any, e, props.onChange);
     };
 
+    const handleCompositionStart = () => {
+      setInputLock(true);
+    };
+
+    const handleCompositionEnd = (e: React.CompositionEvent<HTMLTextAreaElement>) => {
```
Comment (afc163) [inline components/input/TextArea.tsx:90]: 这样把输入逻辑搞的太复杂了，我还是倾向于保持简单，依赖原生 input 的逻辑就好，上层做太多会遇到很多边界问题。

Comment (HomyeeKing) [reply components/input/TextArea.tsx:90]: 其实做着做着 我也有这种想法  ，如果保持原生的话 textarea那部分 是不是要revert掉？
另外今天发现[排版的editable text with limited length](https://ant.design/components/typography-cn/#components-typography-demo-interactive)这部分也有类似问题

Comment (HomyeeKing) [reply components/input/TextArea.tsx:90]: 其实逻辑也还好 就加了个composition的锁，主要是class component的props和state 和相关的一些test case很头疼

Comment (afc163) [reply components/input/TextArea.tsx:90]: 增加的复杂度和获得的收益不成正比，后续遇到边界情况改不动，这一块也很难测试。

Comment (HomyeeKing) [reply components/input/TextArea.tsx:90]: 这样吧 我先不动`Input`, 只修改`TextArea`的部分 ，毕竟`TextArea`之前已经做过[特殊case](https://github.com/ant-design/ant-design/blob/6bcb9ba7c4561b5b1b2c1c874961475f9e141166/components/input/TextArea.tsx#L114)的处理，而且最近的一些issue都基于它

**Pull request closed without merging**